File: Heap/LC_295.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.05

class MedianFinder:

    def __init__(self):
        self.left, self.right = [], [] # left is maxheap, right is minheap
    def addNum(self, num: int) -> None:
        if not self.left or num <= -self.left[0]:
            heapq.heappush(self.left, -num)
        else:
            heapq.heappush(self.right, num)
            
        if len(self.left) - len(self.right) == 2:
            maxval = -heapq.heappop(self.left)
            heapq.heappush(self.right, maxval)
        elif len(self.left) < len(self.right):
            minval = heapq.heappop(self.right)
            heapq.heappush(self.left, -minval)

    def findMedian(self) -> float:
        if len(self.left) > len(self.right):
            return -self.left[0]
        else:
            return (-self.left[0] + self.right[0]) / 2

File: Heap/LC_295.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.05

class MedianFinder {
public:
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if (left.size() == 0 || num <= left.top()){
            left.push(num);
        }
        else{
            right.push(num);
        }
        // maintain the balance of two heaps
        if (left.size() - right.size() == 2){
            int to_pop = left.top();
            left.pop();
            right.push(to_pop);
        }

        if (left.size() < right.size()){
            int to_pop = right.top();
            right.pop();
            left.push(to_pop);
        }
    }
    
    double findMedian() {
        if (left.size() > right.size()){
            return left.top();
        }
        else{
            return (left.top() + right.top()) / 2.0;
        }
    }
private:
    priority_queue<int, vector<int>> left;
    priority_queue<int, vector<int>, greater<int>> right;
    
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */